File: Javascript/Quick-Sort.js
var partition = function(array, p, r) {
  var q = p;
  for (var j = p; r > j; j++) {
    if (array[r] >= array[j]) {
      //swap the elements
      var temp = array[j];
      array[j] = array[q];
      array[q] = temp;
      q++;
    }
  }
  //swap elements
  var temp = array[r];
  array[r] = array[q];
  array[q] = temp;
  //return the index of the pivot
  return q;
}

var quicksort = function(array, l, r) {
  if (l < r) {
    var p = partition(array, l , r); //this is the pivot
    quicksort(array, l, p-1);
    quicksort(array, p+1, r);
  }
}

File: Python/Quicksort.py
#in progress

#partition
def partition(array, p, r):
	q = p
	for j in range(p, r):
    	if array[j] <= array[r]:
        	#swap
        	temp = array[j]
        	array[j] = array[q]
        	array[q] = temp
            q++
    #swap
    temp = array[q]
    array[q] = array[r]
    array[r] = temp
    return q

#quicksort
def quicksort(array, p, r):
	if p < r:
    	q = partition(array, p, r)
        quicksort(array, p, q-1)
        quicksort(array, q+1, r)

#an example
myArray = [0, 2, 1, -100, 0.2, -3.9, 2398, 382]
print(myArray) #array before sorted
quicksort(myArray, len(myArray))
print(myArray) #array after sorted
